ssl/quic/qlog.c

```diff
@@ -0,0 +1,701 @@
+/*
+ * Copyright 2023 The OpenSSL Project Authors. All Rights Reserved.
+ *
+ * Licensed under the Apache License 2.0 (the "License").  You may not use
+ * this file except in compliance with the License.  You can obtain a copy
+ * in the file LICENSE in the source distribution or at
+ * https://www.openssl.org/source/license.html
+ */
+
+#include "internal/qlog.h"
+#include "internal/json_enc.h"
+#include "internal/common.h"
+#include "internal/cryptlib.h"
+
+# define BITS_PER_WORD (sizeof(size_t) * 8)
+# define NUM_ENABLED_W ((QLOG_EVENT_TYPE_NUM + BITS_PER_WORD - 1) / BITS_PER_WORD)
+
+static ossl_unused ossl_inline int bit_get(const size_t *p, uint32_t bit_no)
+{
+    return p[bit_no / BITS_PER_WORD] & (((size_t)1) << (bit_no % BITS_PER_WORD));
+}
+
+static ossl_unused ossl_inline void bit_set(size_t *p, uint32_t bit_no, int enable)
+{
+    size_t mask = (((size_t)1) << (bit_no % BITS_PER_WORD));
+
+    if (enable)
+        p[bit_no / BITS_PER_WORD] |= mask;
+    else
+        p[bit_no / BITS_PER_WORD] &= mask;
+}
+
+struct qlog_st {
+    QLOG_TRACE_INFO info;
+
+    BIO             *bio;
+    size_t          enabled[NUM_ENABLED_W];
+    uint32_t        event_type;
+    const char      *event_cat, *event_name, *event_combined_name;
+    OSSL_TIME       event_time, prev_event_time;
+    OSSL_JSON_ENC   json;
+    int             header_done, first_event_done;
+};
+
+static OSSL_TIME default_now(void *arg)
+{
+    return ossl_time_now();
```

Comment:
The general architectural pattern in all of our QUIC code is to ensure that the time retrieval function can be changed dynamically. This is intended to support testing. So basically any module I write has the time retrieval function as a function pointer.

It's true that QLOG has different options for encoding timestamps, but that would be handled internally by our QLOG code if we wanted to support it, so that's not what motivates this. (We use the relative encoding, see qlog_event_epilogue).